### ISSUE TITLE
Executor rewrite

### ISSUE THREAD

**Issue by RiscInside** [RiscInside/chronos#1]
This is a tracking issue for executor rewrite.

### Why?

The current version of the executor is "fine" for CPU-intensive applications, but it has a few limitations
1. Use of signals to pause/resume threads prevents as from pausing them in kernel space
2. CPUs used in the experiment are 100% hijacked from the main system, making the system less usable (or not usable at all on single-core systems)
3. Currently executor does not distinguish between thread waiting on a lock and CPU-intensive work, which means that the executor takes more time to execute than it actually should

### What can we do?

- [x] The first issue can be addressed by using a different primitive to suspend threads. Currently, I am considering using barrier thread with a low priority. Tasks to be suspended will be assigned an even lower priority, ensuring that they will never (or will rarely, more on that later) be scheduled to run. Since the Linux scheduler does not treat kernel code differently for the purposes of scheduling, we should be able to pause code running in kernel space.
- [ ] Next, the CPU usage issue can be used by occasionally suspending the entire scheduler. Linux kernel can do this for us (see https://man7.org/linux/man-pages/man7/sched.7.html, the section titled "Limiting the CPU usage of real-time and deadline processes"). We also need to allow tasks to run on the scheduler CPU as well, so we don't have one CPU lying around. This should allow the executor to run single-core systems while keeping the system responsive. A good test case for that would be running a multithreaded program that uses all available CPU cores.
- [ ]  Finally, the last issue can be addressed by instrumenting all system calls that make thread interact in "suspend/resume" ways.

#### Thread suspension

At the time of writing, I am aware of three possible ways thread can be unwillingly suspended/resumed on Linux.

1. `ptrace(PTRACE_INTERRUPT)` and `ptrace(PTRACE_CONT)`. System call either completes or gets interrupted
2. `tgkill(tid, SIGSTOP)` and `tgkill(tid, SIGCONT)`. Same exact problem
3. Barrier method, see below

The barrier method works by having a thread that keeps the CPU busy. Once we want to suspend a process, we set its affinity to run on the same CPU barrier runs on and its priority to be lower than that of the barrier. Depending on our choice of policies, this could either prevent the thread from running completely or give suspended *very* little time to progress.

There is no question about the scheduler running with `SCHED_FIFO`. All scheduler operations are constant-time and the scheduler only runs once in a while, so this shouldn't be a problem.

Assume that the scheduler thread has the `SCHED_FIFO` priority of `N` (we might want to have it lower than the maximum to allow some more "legitimate" real-time processes to run). Runnable thread can then run with a real-time priority of  `N - 1`, ensuring that it will be preempted every time scheduler wakes up. This also gives us a way to "disable interrupts" by elevating the priority of the running thread back to `N`. This will be extremely useful for `futex()` instrumentation, more on that later.

There are two options for the barrier thread
1. Run it at a real-time priority of `M` < `N - 1` and run suspended threads at real-time priority `M - 1`. This is a "hard freeze", ensuring that the suspended threads don't get a chance to run. The downside, of course, is that the CPU barrier runs on effectively becomes useless for the runtime of the experiment (unless a thread is running on it), This should not make the system completely unusable, however, as there is a ceiling on how much time real-time processes can run for combined.
2. Run it at a low priority within the `SCHED_OTHER` or `SCHED_BATCH` class. Suspended threads will then be assigned the `SCHED_IDLE` scheduling class. This would allow other processes to run on the system, at the expense of this only being a soft freeze: `SCHED_IDLE` threads tend to periodically get scheduled, although they don't tend to be scheduled for long. One way to mitigate that would be looking at how much time idle threads actually ran for an account for that while calculating "vruntime" when they get scheduled again.

It seems that 1st approach is slightly easier to implement and it offers more accurate simulation, so I will roll with that for now.

#### Instrumenting thread interactions.

Imagine the following scenario
1. Thread no. 1 goes to sleep for some reason, waiting for thread no. 2 (which may be unknown at the time)
2. Thread no. 2 does some work and signals thread no. 1 to wake up.
3. Thread no. 1 gets woken up by thread no. 2

This happens all the time (locks, joins, signals, etc). Ideally, we would like to instrument all such situations. Here is the way we can do that:
- [ ] Add a method that allows us to detach from runtime for a bit. This can work as follows:
1. Thread elevates its priority to that of the scheduler. Hopefully, this is not too expensive of an operation, as runnable threads in real-time class should not cause any rescheduling by increasing their own priority
2. Set some flag politely asking the scheduler to not schedule us again. The scheduler will be able to see that the flag is set once it gets to run after we finish (or once the scheduler exits nanosleep())
- [ ] Use eBPF to implement vruntime communication between the wakee and the waker. This can work as follows:
1. Sleepers use the bpf() syscall to update `wakeup_vruntime` BPF map: `wakeup_vruntime[tid] = current_vruntime()`. This is not technically necessary, but it allows us to ensure that 1) BPF map update won't fail from within the tracepoint and 2) we can detect situations where a thread that is non-negligibly behind in vruntime wakes up a thread that is ahead in vruntime (this shouldn't happen).
2. Wakers do the same thing. They do not need to elevate their priorities, however.
3. Finally, we instrument the `tp/sched/sched_wakeup` tracepoint. It needs to 1) recalculate vruntime (this is probably the hardest part since we would need to communicate the vruntime to the userspace somehow, so we would have to touch userspace memory)
- [ ] We also need to implement a method that would allow threads to attach back to the runtime. It should accept the new vruntime as an argument.
A good demo would be having a thread scale down its tdf to 0.1, and then do some work, while having other threads wait on the first one with `pthread_join` (which uses one of the system calls from the wait family). In the future, we might also want to instrument sleep() system call for the same reason.
